### ISSUE TITLE
Add `aroundEach` Hook Similar to RSpec `around`

### ISSUE THREAD

**Issue by klondikemarlen** [vitest-dev/vitest#5728]
### Clear and concise description of the problem

I want an `aroundEach` so that I can run my database tests inside a transaction and roll back the transaction after the test completes. This would _vastly_ speed up my test runs.

Currently I need to wipe each table in my database, on each test run, to be sure that test state is clean.
Naturally, due to having a great many migrations, this is more performant than dropping the whole database and then rebuilding the schema.

It's also possible that this is already possible, but after spending an entire day trying to get it to work, I'm at least _reasonably_ sure it isn't.

### Suggested solution

Add support for a simple `aroundEach` hook.

```ts
// sequelize being a sequelize 7 (or 6) instance that is using managed transactions

aroundEach(async (example) => {
   await sequelize.transaction(async (transaction) => {
     await example.run()
     await transaction.rollback()
   })
})
```

This would vastly speed from something like
```
 tests/services/position-teams/create-service.test.ts (3) 1968ms
   ✓ api/src/services/position-teams/create-service.ts (3) 1967ms
     ✓ CreateService (3) 1967ms
       ✓ #perform (3) 1967ms
         ✓ when passed valid attributes, it creates a position team 770ms
         ✓ when teamId is missing, errors informatively 614ms
         ✓ when positionId is missing, errors informatively 582ms
```
To something much more reasonable.

Its true that I can do this in each test, but who wants to clutter up their test code with setup logic?

### Additional context

Will be used in many of the projects I'm working on starting with https://github.com/icefoganalytics/wrap (currently not on `main` but using Vitest in https://github.com/icefoganalytics/wrap/pull/32/files)

### Validations

- [X] Follow our [Code of Conduct](https://github.com/vitest-dev/vitest/blob/main/CODE_OF_CONDUCT.md)
- [X] Read the [Contributing Guidelines](https://github.com/vitest-dev/vitest/blob/main/CONTRIBUTING.md).
- [X] Read the [docs](https://vitest.dev/guide/).
- [X] Check that there isn't already an issue that request the same feature to avoid creating a duplicate.

**Comment by hi-ogawa** [comment]
I think I get the idea, but can you also provide more concrete code and references? For example, what is "RSpec `around`"? also I would assume you're using `sequelize.transaction` with async hook mode https://sequelize.org/docs/v6/other-topics/transactions/#automatically-pass-transactions-to-all-queries but it's not entirely clear from the given code.

We are not necessary familiar with all those stuff, so providing as much context as possible would help triage the request better without misunderstanding.

**Comment by hi-ogawa** [comment]
I think what people do so far to wrap a whole test function is to either create an own wrapper of `it` like `effect` https://github.com/Effect-TS/effect/blob/40c2b1d9234bcfc9ab4039282b621ca092f800cd/packages/vitest/src/index.ts#L55-L64 or I think it's also possible to use custom task function https://vitest.dev/advanced/runner.html#your-task-function

**Comment by klondikemarlen** [comment]
> I think I get the idea, but can you also provide more concrete code and references? For example, what is "RSpec `around`"? also I would assume you're using `sequelize.transaction` with async hook mode https://sequelize.org/docs/v6/other-topics/transactions/#automatically-pass-transactions-to-all-queries but it's not entirely clear from the given code.
> 
> We are not necessary familiar with all those stuff, so providing as much context as possible would help triage the request better without misunderstanding.

Thanks for the suggestions! I've updated the request to include some references. Also realized that the repo links I shared are still private (~they'll be public soon, just hasn't happened yet~ may never happen 😞). I'll shared some code examples directly.

**Comment by klondikemarlen** [comment]
> I think what people do so far to wrap a whole test function is to either create an own wrapper of `it` like `effect` https://github.com/Effect-TS/effect/blob/40c2b1d9234bcfc9ab4039282b621ca092f800cd/packages/vitest/src/index.ts#L55-L64 or I think it's also possible to use custom task function https://vitest.dev/advanced/runner.html#your-task-function

I'll see if I can make this work.
Maybe I can add a `describe({ type: 'database' }, () => {}` or something that using that effect pattern you shared.

**Comment by sheremet-va** [comment]
Last year I gave this example: https://github.com/vitest-dev/vitest/issues/3404#issuecomment-1554230846

Since then, the API has changed a little bit:

```ts
import { createTaskCollector, getCurrentSuite } from 'vitest/suite'

export const withTransaction = createTaskCollector(
  function (name, fn, timeout) {
    const handler = async (...args) => {
      await database.transaction(() => fn(...args))
    }
    getCurrentSuite().task(name, {
      ...this, // so "todo"/"skip"/... is tracked correctly
      handler,
      timeout,
    })
  }
)
```

**Comment by sheremet-va** [comment]
Maybe you can also do something like this:

```ts
import { test as baseTest } from 'vitest'

export const withTransaction = baseTest.extend({
  transaction: ({ database }, use) => {
    await database.transaction(() => {
      await use()
    })
  }
})
```

I haven't tried it, but I think it should work 🤔

**Comment by klondikemarlen** [comment]
That looks like a _huge_ step closer to what I was trying do to, though I don't think it quite solves the usability issue. The issue in question is: If the the next developers on the project forget to use the transaction helper, it will lead to hard to understand bugs. Ideally the everyday developer wouldn't have to worry about database cleanup as it should be handled at the config level (i.e https://vitest.dev/config/#setupfiles).

Given that, I'm going to see if I can implement an `aroundEach`, using the task collector? (or something), so database cleanup is handled at the highest level. For context I'm writing tests for the back-end of a web app, so 90% of tests will be database tests. I'd like to operate so that tests run the database cleaner by default, and if there are some non-database tests, that we want maximum speed for, we can set a flag to disable them.

e.g.
```ts
describe(() => {
    // some database test
})
```
Would always run in a transaction, and auto-rollback after completion.

If we have a library or utility test that doesn't use the database we could turn the database cleaner off via
```ts
describe({ type: "library" }, () => {
   // some test that doesn't required the database
})
```
Or whatever vitest supports.

**Comment by klondikemarlen** [comment]
After all my efforts and this probably? working `aroundEach`
```ts
async function aroundEach(fn: (runExample: () => Promise<void>) => void) {
  beforeEach(async () => {
    let resolveRunExample: (value: void | PromiseLike<void>) => void
    const runExample = new Promise<void>((resolve) => {
      resolveRunExample = resolve
    })

    fn(() => runExample)

    return async () => {
      resolveRunExample()
      await runExample
    }
  })
}
```
The following code _does not_ pass the transaction to the various Sequelize model actions performed in the "runExample()".
```ts
aroundEach(async (runExample) => {
  try {
    await sequelize.transaction(async () => {
      await runExample()
      return Promise.reject("TRIGGER DATABASE CLEANUP")
    })
  } catch (error) {
    if (error !== "TRIGGER DATABASE CLEANUP") {
      throw error
    }
  }
})
```

**Comment by klondikemarlen** [comment]
Unless anyone still thinks this feature is valuable, I'm going to close it, as my primary use case fails for unrelated reasons.

**Comment by mdnorman** [comment]
> Maybe you can also do something like this:
> 
> 
> 
> ```ts
> 
> import { test as baseTest } from 'vitest'
> 
> 
> 
> export const withTransaction = baseTest.extend({
> 
>   transaction: ({ database }, use) => {
> 
>     await database.transaction(() => {
> 
>       await use()
> 
>     })
> 
>   }
> 
> })
> 
> ```
> 
> 
> 
> I haven't tried it, but I think it should work 🤔 

I was hoping to be able to use a fixture for something like this, but apparently the async context doesn't get passed to the underlying test function. See https://github.com/vitest-dev/vitest/issues/5858

**Comment by trim21** [comment]
so the current solution to do this is create a helper function and wrap it with all current and future tests?

**Comment by klondikemarlen** [comment]
@trim21 So ... I haven't been able to get transaction wrapping with rollback to work in a nice way, so I'm currently doing this kind of gross thing https://github.com/icefoganalytics/wrap/blob/main/api/tests/setup.ts for MSSQL.
```ts
/**
 * See https://vitest.dev/config/#setupfiles
 *
 * Run some code before each test file.
 *
 * WARNING: Be very careful of imports in this file!!!
 * Vitest will not mock modules that were imported inside a setup file because they are
 * cached by the time a test file is running.
 * You can do
 * ```ts
 * vi.hoisted(() => {
 *   vi.resetModules()
 * })
 * ```
 * to clear all module caches before running a test file.
 * See: https://vitest.dev/api/vi#vi-mock
 */

import dbLegacy from "@/db/db-client-legacy"
import { logger } from "@/utils/logger"

import * as matchers from "@/support/matchers"

// Add New Matchers
expect.extend(matchers)

async function getTableNames() {
  const query = /* sql */ `
    SELECT
      table_name AS "tableName"
    FROM
      information_schema.tables
    WHERE
      table_schema = 'dbo'
      AND table_type = 'BASE TABLE'
      AND table_name NOT IN ('knex_migrations' , 'knex_migrations_lock');
  `

  try {
    const result = await dbLegacy.raw<{ tableName: string }[]>(query)
    const tableNames = result.map((row) => row.tableName)
    return tableNames
  } catch (error) {
    logger.error("Error fetching table names: " + error)
    throw error
  }
}

async function getTableNamesWithoutIdentityColumn() {
  const query = /* sql */ `
    SELECT
      tables.name AS "tableName"
    FROM
      sys.tables AS tables
    WHERE
      tables.schema_id = SCHEMA_ID('dbo')
      AND NOT EXISTS (
        SELECT 1
        FROM sys.columns AS columns
        WHERE columns.object_id = tables.object_id AND columns.is_identity = 1
      );
  `

  try {
    const result = await dbLegacy.raw<{ tableName: string }[]>(query)
    const tableNames = result.map((row) => row.tableName)
    return tableNames
  } catch (error) {
    logger.error("Error fetching table names without identity columns: " + error)
    throw error
  }
}

/**
 * Example of generated SQL commands used for cleaning database:
 *
 * ```sql
 * ALTER TABLE table1 NOCHECK CONSTRAINT ALL;
 * ALTER TABLE table2 NOCHECK CONSTRAINT ALL;
 * ...
 * DELETE FROM table1 WHERE 1=1;
 * DELETE FROM table2 WHERE 1=1;
 * ...
 * DBCC CHECKIDENT ('table1', RESEED, 0);
 * DBCC CHECKIDENT ('table2', RESEED, 0);
 * ...
 * ALTER TABLE table1 CHECK CONSTRAINT ALL;
 * ALTER TABLE table2 CHECK CONSTRAINT ALL;
 * ...
 * ```
 */
async function buildCleanupQuery() {
  const tableNames = await getTableNames()
  const tableNamesWithoutIdentityColumn = await getTableNamesWithoutIdentityColumn()
  const tableNamesWithIdentityColumn = tableNames.filter(
    (tableName) => !tableNamesWithoutIdentityColumn.includes(tableName)
  )
  const disableAllConstraintsQuery = tableNames
    .map((tableName) => /* sql */ `ALTER TABLE ${tableName} NOCHECK CONSTRAINT ALL;`)
    .join("\n")
  const deleteAllInAllTablesQuery = tableNames
    .map((tableName) => /* sql */ `DELETE FROM ${tableName} WHERE 1=1;`)
    .join("\n")
  const resetIdentityColumnsQuery = tableNamesWithIdentityColumn
    .map((tableName) => /* sql */ `DBCC CHECKIDENT ('${tableName}', RESEED, 0);`)
    .join("\n")
  const enableAllConstraintsQuery = tableNames
    .map((tableName) => /* sql */ `ALTER TABLE ${tableName} CHECK CONSTRAINT ALL;`)
    .join("\n")
  const cleanupQuery = [
    disableAllConstraintsQuery,
    deleteAllInAllTablesQuery,
    resetIdentityColumnsQuery,
    enableAllConstraintsQuery,
  ].join("\n")
  return cleanupQuery
}

async function cleanDatabase() {
  const cleanupQuery = await buildCleanupQuery()
  try {
    await dbLegacy.raw(cleanupQuery).catch(console.error)
    return true
  } catch (error) {
    logger.error(error)
    return false
  }
}

beforeEach(async () => {
  await cleanDatabase()
})
```

The blocker for me was that Sequelize 7 use of [AsyncLocalStorage](https://nodejs.org/api/async_context.html#class-asynclocalstorage), loses context with the `aroundEach` version I was using, and doesn't detect the transaction.

I'm not sure if its a Sequelize problem, or and AsyncLocalStorage problem, but I'm pretty sure that the problem isn't with Vitest. Other than the fact that Vitest doesn't provide a convenient `aroundEach`.

**Comment by manvydas-tbol** [comment]
Would be nice to have this implemented. We're using a global config, which we would like to adjust for some tests, but we can't, since asyncLocalStorage doesn't work and there is no way to access vitest tests' context outside of the test's local code itself.

**Comment by fragkakis** [comment]
Would live to see this. This is what Rails does, and testing is a bliss.

**Comment by gabrielDonnantuoni** [comment]
Would love to see this too. My use case would be wrapping with an AsyncLocalStorage

**Comment by ryant26** [comment]
Similarly, I would like to have this for use with AsyncLocalStorage

**Comment by lbargery** [comment]
Would also like this for async local storage! (Db transactions and some global logging bits)

**Comment by gabrielDonnantuoni** [comment]
For those who would use it with AsyncLocalStorage. I'm using this approach as workaround.

There's a module '@/storage' that exports a function `useStore` that returns store.getStore(). 
All modules that depends on the storage import from '@/storage'

So I built a mock factory that overwrite `useStore` and has an usage that looks like this
```ts
 // Mock useStore with default value
 vi.mock('@/storage', buildStoreMockFactory())
 // or with custom overwrite
 vi.mock('@/storage', buildStoreMockFactory({
   var: {
     lang: Lang.EN_US,
   },
 }))
```

**Comment by rossipedia** [comment]
For `AsyncLocalStorage` at least, won't something like this do the trick?

```ts
const storage = new AsyncLocalStorage();

beforeEach(() => {
  storage.enterWith(someValue);
});

test('foo', async () => {
  expect(storage.getStore()).not.toBeUndefined();
});

// etc...
```

**Comment by lbargery** [comment]
@rossipedia Sadly that doesn't work since the async context isn't persisted as the same context between beforeEach and test. Either vittest would need to guarantee that or you need the aroundEach (I think!)

**Comment by rossipedia** [comment]
> @rossipedia Sadly that doesn't work since the async context isn't persisted as the same context between beforeEach and test. Either vittest would need to guarantee that or you need the aroundEach (I think!)

I'll have to do some more digging. It seems to work when I try it, but it that might be a side effect rather than an intentional contract.

**Comment by lbargery** [comment]
I think if you turn isolation false so tests run in parallel you'll find issues (which for my use anyway was the key thing I wanted)

**Comment by rossipedia** [comment]
Actually, running with `--no-isolate` didn't cause issues, but what did was running the `beforeEach` that called `AsyncLocalStorage.enterWith()` inside a `describe` block. `getStore()` just returned `undefined` in that scenario

It works when the `beforeEach` is at module scope, just not when nested. This leads me to think the "working" behavior is accidental.

It would definitely be nice to have some way of wrapping the actual test call with an `AsyncLocalStorage.run()` or similar :+1:

**Comment by dimfeld** [comment]
There's some difference here between Node 22 and Node 24 as well. Calling `enterWith` in a beforeEach worked for me in Node 22, but the same code in Node 24 fails because it can't find the context inside the test.

I assume this is related to Node 24 switching to `AsyncContextFrame` which has been found in other issues to treat `enterWith` slightly differently. Don't have a solution yet but if I come up with something better than just wrapping everything in a `with` callback I'll report back.

Edit: I ended up using `vi.spyOn`. It's not a great solution and won't work in all cases (e.g. concurrent tests in a file or additional calls to `run` in the tested code) but works for my purposes.

```js
beforeEach((ctx) => {
  const context = {  value: "test context" };
  const storeMock = vi
    .spyOn(serviceContextModule.__testing.store, 'getStore')
    .mockReturnValue(context);

  ctx.onTestFinished(() => {
    storeMock.mockRestore();
  });
});
```

**Comment by chriswheeldon** [comment]
I've been digging into this a bit as I am also interested. I'm using node `v24.11.0` and a local checkout of vitest @ 06d6207fb76830d029acf570e67814f97699dbe2.

I'm sharing my findings here on propagating the async context as they may be useful.

For me, the fixture approach with `asyncLocalStorage.enterWith` does work out of the box. I believe the only caveats are,
1. The fixture must be the first fixture to be resolved [here](https://github.com/vitest-dev/vitest/blob/9a79b90c9543f896aba2d304873305d33c19fc65/packages/runner/src/fixture.ts#L176).
2. The fixture must call `enterWith` before any other asynchronous call.
3. There must be no describe scoped `beforeEach` blocks that use the fixture **(see end of post for more details and what I think is a workaround)**

If those conditions are not met then the `enterWith` will not be called within the context that (eventually) runs the test itself.

Here's an example that works for me:

```ts
import { AsyncLocalStorage } from 'node:async_hooks'
import { describe, expect, test } from 'vitest'

const asyncLocalStorage = new AsyncLocalStorage<{ value: string }>()

const withAsyncLocalStorage = test.extend({
  als: async ({}, use) => {
    asyncLocalStorage.enterWith({ value: 'from-withContext' })
    await use({})
  },
})

describe('Debug AsyncLocalStorage with withContext', () => {
  withAsyncLocalStorage('check if context persists in withContext test', ({ als }) => {
    const store = asyncLocalStorage.getStore()
    expect(store).toBeDefined()
    expect(store?.value).toBe('from-withContext')
  })
})
```

However, I've found that attempting to use the async context in a `beforeEach` block inside a `describe` will mean that it is no longer available in the test block itself i.e. adding the following will cause the test to fail (n.b. the `beforeEach` itself will execute successfully).
```ts
describe(...
  beforeEach(({ als }) => {
    const store = asyncLocalStorage.getStore()
    expect(store).toBeDefined()
    expect(store?.value).toBe('from-withContext')
  })

  ...
```

I think this is because the introduction of the fixture in the describe scoped `beforeEach` will cause the fixture setup code to be executed as part of the `beforeEach` execution and such `beforeEach` blocks are not currently executed in the async context that will execute the test itself. This is because vitest [always attempts to call the parent suite hooks first](https://github.com/vitest-dev/vitest/blob/1aac59cd207829dfa42ce7f48c82e1e8a6935dbe/packages/runner/src/run.ts#L140-L144) (regardless of whether there are any such hooks or not) and then call the suite level beforeEach blocks. This has the effect of moving the suite hooks into their own async context.

A workaround is to define a top level `beforeEach` block to force execution of the fixture setup in the outer async context. This will then be correctly propagated into describe scoped `beforeEach`s and the test itself i.e.

```ts
import { AsyncLocalStorage } from 'node:async_hooks'
import { beforeEach, describe, expect, test } from 'vitest'

const asyncLocalStorage = new AsyncLocalStorage<{ value: string }>()

const withAsyncLocalStorage = test.extend({
  als: async ({}, use) => {
    asyncLocalStorage.enterWith({ value: 'from-withContext' })
    await use({})
  },
})

beforeEach(({ als }) => {
  const store = asyncLocalStorage.getStore()
  expect(store).toBeDefined()
  expect(store?.value).toBe('from-withContext')
})

describe('Debug AsyncLocalStorage with withContext', () => {
  beforeEach(({ als }) => {
    const store = asyncLocalStorage.getStore()
    expect(store).toBeDefined()
    expect(store?.value).toBe('from-withContext')
  })

  withAsyncLocalStorage('check if context persists in withContext test', ({ als }) => {
    const store = asyncLocalStorage.getStore()
    expect(store).toBeDefined()
    expect(store?.value).toBe('from-withContext')
  })
})
```

**Comment by chriswheeldon** [comment]
Update to the above: the use of a fixture is not required and `enterWith` can be used successfully in a `beforeEach` block provided it is the first `beforeEach` block to run and it is defined at file level.

**Comment by rossipedia** [comment]
> Update to the above: the use of a fixture is not required and `enterWith` can be used successfully in a `beforeEach` block provided it is the first `beforeEach` block to run and it is defined at file level.

That's what I found as well, which (IMO) feels a bit too brittle to rely on.

**Comment by chriswheeldon** [comment]
@rossipedia, my bad, I was so lost in the fixture idea that I hadn't fully absorbed what you wrote [here](https://github.com/vitest-dev/vitest/issues/5728#issuecomment-3448856723).

Just to ask the obvious question, though I understand it might feel risky to commit to, but @sheremet-va would you be willing for vitest to guarantee (in the form of a test) that the async context of the first file level `beforeEach` block is the same as the test execution?

**Comment by lbargery** [comment]
My interest was using an async context with isolation: false, so you have multiple tests running at one time. Does the above work in that situation? If I find time I can try to play myself too. Thanks for investigating so far

**Comment by calirvine** [comment]
AI and I have been playing around with this idea and I think I prefer an API closer to `test.extend` over `aroundEach` (though maybe both can be supported). The API I have been leaning towards is 
`describe.wrap(fn, {scope: 'suite' | 'test'}` and `test.wrap(fn)`

It's still rough and largely AI generated, so take it with a grain of salt, but thought I'd share the draft in case it gets any creative ideas flowing. 
https://github.com/vitest-dev/vitest/compare/main...calirvine:vitest:feat/wrappers-wrap-api

**Comment by klondikemarlen** [comment]
@calirvine That looks interesting!

What I really want is this interface from RSpec https://blog.appsignal.com/2024/02/07/a-deep-dive-into-rspec-tests-in-ruby-on-rails.html

e.g.
```ruby
RSpec.describe 'some tests that use the database', type: :db do
    # tests are wrapped in a transaction
end
```
vs.
```ruby
RSpec.describe "some tests that don't use the database", type: :lib do
  # tests skip the database transaction rollback processs
end
```

With an ability to configure the default "type" so that
```ruby
RSpec.describe 'tests with no "type" are assumed to be database tests' do
    # tests are wrapped in a transaction
end
```

Specifically I want this Ruby tool in JS 😉 https://github.com/DatabaseCleaner/database_cleaner?tab=readme-ov-file#rspec-example

So I can do something like
```ruby
RSpec.configure do |config|
  config.before(:suite) do
    DatabaseCleaner.strategy = :transaction
    DatabaseCleaner.clean_with(:truncation)
  end

  config.around(:each) do |example|
    DatabaseCleaner.cleaning do
      example.run
    end
  end
end
```

I'd even be tempted to build it (at least so it works with Sequelize) ... but I haven't been able to determine a user friendly way to configure it in Vitest due to the loss of AsyncLocalStorage context in an "aroundEach" like effect.

EDIT: I think https://rspec.info/features/6-0/rspec-rails/directory-structure/ has some interesting examples about why this is useful. Particularly because you can do stuff like
```ruby
​# set `:type` for serializers directory
RSpec.configure do |config|
  config.define_derived_metadata(:file_path => Regexp.new('/spec/serializers/')) do |metadata|
    metadata[:type] = :serializer
  end
end
```
And effectively have a different set of default actions that you are running or features that you are including on a per-folder basis.

**Comment by calirvine** [comment]
@klondikemarlen testing in rails is my inspiration as well. With this approach you'd be able to do something along the lines of:

```ts
const describe = baseDescribe.wrap(async (run) => {
  db.transaction(() => {
    run()
  });
}, {scope: 'suite'});

const test = baseTest.wrap(async (run) => {
  db.transaction(() => {
    run()
  });
});

describe('example', async () => {
  beforeAll(() => {
    // load fixtures for all tests here, we are in a suite scoped transaction that will be cleaned up after the suite
  });

  test('example test', async () => {
    // We are in a nested transaction here, do whatever you need to and it will be rolled back to suite transaction state between tests
  });
});
```

But I'd need to flesh this out more and validate everything still. Just sketching at this point.

**Comment by sheremet-va** [comment]
`aroundEach` and `aroundAll` are going to be released as part of Vitest 4.1 (in the next beta 3). The docs are already available [here](https://deploy-preview-9450--vitest-dev.netlify.app/api/hooks.html#aroundeach).

```ts
import { aroundEach, test as base } from 'vitest'

const test = base.extend<{ db: Database }>({
  db: async ({}, use) => {
    // db is created before `aroundEach` hook
    const db = await createTestDatabase()
    await use(db)
    await db.close()
  },
})

test.aroundEach(async (runTest, { db }) => {
  await db.transaction(runTest)
})

test('insert user', async ({ db }) => {
  // called inside a transaction
  await db.insert({ name: 'Alice' })
})
```
